swapmyvote/swapmyvote
# issue 772 tv advice party links

Addresses and closes #772 Relate Tactical Vote recommendations to Parties in database

This unblocks #773 Annotate Swaps with parties in tactical vote recommendations

Built on the PR for #768, best to review that first.

# Summary of approach

New table created RecommendedParty where rows represent what one TV advisor said about one party in one constituency.

3 current TV imports were re-written to discover actual parties that the advice text recommended.

Where the text advice implies multiple parties, multiple records are created.

# Steps:

- better name: Party#smv_code
- add rails model to link parties to TV recommendations
- make recommended parties findable from constituency
- make recommendations findable from party
- method for parties standing in constituency
- ensure dont recommend Plaid outside Wales, etc
- total re-write of STT import
- extract method for recommended parties update
- total re-write of sprintforpr import
- total re-write of tactical.vote import

# To Verify

- check out branch
- get db up to date with either

    rake db:schema:load

  or

    rake db:migrate

- seed the database

    ELECTION_TYPE=g rake db:seed

- use rails console to explore

    \# all recommended parties for one constituency
    OnsConstituency.find_by_name("Torfaen").recommended_parties.all.map { |rp| [rp.site, rp.party.name] }

    \# Five constituencies where labour is recommended
    labour = Party.find_by_smv_code(:lab)
    RecommendedParty.joins(:constituency).where(party_id: labour.id ).limit(5).all.map { |rp| [rp.constituency.name, rp.party.name] }
